THRIFT-3873: fix compiler warnings on windows with VS2010

jeking3 opened the pull request:
Also changed VERSION to THRIFT_VERSION to avoid conflicts with third party or OS headers.

ben-craig commented on compiler/cpp/CMakeLists.txt:
```diff
@@ -18,6 +18,16 @@
+    endif(HAVE_STDINT_H)
+
+    # The generated code also causes a size type warning we cannot work around easily
+    set_source_files_properties(thriftl.cc PROPERTIES COMPILE_FLAGS /wd4267)
```
This should be resolved by putting a pragma in the source code.  Only put code relevant things in the build system if there are no other alternatives.

jeking3 replied:
I was able to move this into thriftc.ll however we still need the /FIstdint.h to avoid compiler warnings.

ben-craig commented on lib/cpp/src/thrift/transport/THeaderTransport.h:
```diff
@@ -135,8 +141,9 @@ class THeaderTransport : public TVirtualTransport<THeaderTransport, TFramedTrans
-    int trans = writeTrans_.size();
-    return trans;
+    std::vector<uint16_t>::size_type trans = writeTrans_.size();
+    assert(trans <= UINT16_MAX);
```
If your casting error checking is just debug asserts, then you might as well stick with the casts.
My preference would be to throw an exception if the size is out of bounds.

jeking3 replied:
I followed the pattern that already exists in other implementation files like THeaderTransport.cpp.  I agree with you here, I think it would be better to have code like:

```
if (trans > std::numeric_limits<uint16_t>().max()) { 
  throw TTransportException(CORRUPTED_DATA, ...); 
}
```

I think it would be better to open a new story to do this conversion across the codebase.  There have to be at least 10 of these checks in THeaderTransport.cpp alone.  Can we leave these as-is and address them holistically?

I did change the ones in the compiler code to use the existing pattern of throwing a std::string with an error message when something goes wrong.

ben-craig commented on lib/cpp/src/thrift/transport/TServerSocket.cpp:
```diff
@@ -60,7 +60,7 @@
 
-#if defined(_WIN32) && (_WIN32_WINNT < 0x0600)
+#if defined(_WIN32) && (_WIN32_WINNT < 0x0600) && !defined(AI_ADDRCONFIG)
   #define AI_ADDRCONFIG 0x0400
```
You didn't break this, but this kind of thing is exactly what the platform socket header is for.  Defining constants that don't belong to us often leads to trouble.  Instead, we should be defining something like THRIFT_AI_ADDRCONFIG that can either be the real AI_ADDRCONFIG or 0x0400.

jeking3 replied:
I'm not convinced this needs to be defined here; every windows platform windows supports has winsock2, which includes ws2def.h, which defines this and AI_PASSIVE which is used right alongside AI_ADDRCONFIG in the code... so I am going to remove the #if block here.

jeking3 commented:
@ben-craig Thanks these are great comments, I will make some changes to improve this.

hcorg commented on compiler/cpp/src/thrift/generate/t_erl_generator.cc:
```diff
@@ -968,10 +969,13 @@ void t_erl_generator::export_string(string name, int num) {
-
+  t_struct::members_type::size_type num = tfunction->get_arglist()->get_members().size();
+  if (num > std::numeric_limits<int>().max()) {
+    throw "integer overflow in t_erl_generator::export_types_function, name " + tfunction->get_name();
```
throwing std::string is an antipattern - it should at least be wrapped in std::runtime_error 

this code is also repeated in couple of places, maybe some utility function could simplify it?

something like: 
`template <typename To, typename From> To safe_numeric_cast(From i) { ... }`

jeking3 replied:
Throwing std::string or char * is a pattern used throughout the compiler code.  While I agree with you this is not something that should be done, I was following what was already there for continuity and maintainability.  

I filed https://issues.apache.org/jira/browse/THRIFT-3982 to cover this issue across the compiler codebase.  As such, I would like to request this remain.

hcorg replied:
true... I've forgot about compiler "patterns" :) yeah, we need to fix it globally

hcorg commented on compiler/cpp/src/thrift/generate/t_go_generator.cc:
```diff
@@ -433,7 +434,11 @@ std::string t_go_generator::camelcase(const std::string& value) const {
-      fix_common_initialism(value2, i);
+
+	  if (i > std::numeric_limits<int>().max()) {
+        throw "integer overflow in t_go_generator::camelcase, value = " + value;
```
looks like some indentation problem (some time ago `make format` was helping with such issues... but it might no longer work :/ )

jeking3 replied:
I was using Visual Studio 2010 for many of these changes, which was inserting tabs.  I will clean this up.

hcorg commented on compiler/cpp/src/thrift/thriftl.ll:
```diff
@@ -39,11 +39,17 @@
+// avoid isatty redefinition
 #define YY_NEVER_INTERACTIVE 1
 
+// no <unistd.h>
```
unnecessary comment imho

jeking3 replied:
Removed in next push.

hcorg commented on lib/cpp/src/thrift/transport/THeaderTransport.cpp:
```diff
@@ -375,7 +383,12 @@ void THeaderTransport::resetProtocol() {
+  if (wb > std::numeric_limits<uint32_t>().max()) {
+    throw TTransportException(TTransportException::CORRUPTED_DATA,
+                              "write size is unreasonable");
+  }
```
as before - code repetition - some utility function would make it much cleaner (different than before, as now TTTransportException should be thrown)

jeking3 replied:
I added apache::thrift::transport::safe_numeric_cast to the TTransportException header.  I couldn't think of a better place to put it for now... I agree it simplifies the code in many places, and the transport code seems to be the code closest to the platform where these transform issues exist.

jeking3 replied:
Unfortunately using a template like this caused build warnings to occur when the inputs are not the same signed-ness.  I think these bounds checks need to be done with eyes wide open, and need to be specific in each case in order to avoid compiler warnings.

hcorg commented on lib/cpp/src/thrift/transport/THeaderTransport.cpp:
```diff
@@ -394,15 +411,19 @@ void THeaderTransport::setHeader(const string& key, const string& value) {
     maxWriteHeadersSize += 5 + 5 + (it->first).length() + (it->second).length();
   }
-  return maxWriteHeadersSize;
+  if (maxWriteHeadersSize > std::numeric_limits<uint32_t>().max()) {
```
I've just realized - if somebody will compile this code on 32bit machine with gcc he will get "comparison always true" warning... just a thought

jeking3 replied:
Compiling in Win32 exposed additional warnings I will clean up.  Thanks for this comment!  I am also going to build with Visual Studio 2015 in 32-bit and 64-bit to make sure they are all cleaned up.  I have Win64 VS2010 building without any warnings at this point.  In addition I found that the tutorial CMakeLists.txt assumes ZLIB is configured so I changed it to be conditional on the ZLIB library.

jeking3 replied:
I have built all third party libraries under Visual Studio 2010, 2012, 2013, and 2015 in x86 and x86 mode, in debug and release, in static and dynamic mode, where appropriate.  Then I build thrift in the same configurations.  I will be releasing the windows build instructions and contributed build batch files to the project so that they can be used as a reference on how to build thrift on windows for C++ with all of the features enabled, using cmake.

hcorg commented on lib/cpp/src/thrift/transport/THeaderTransport.cpp:
```diff
@@ -494,7 +519,12 @@ void THeaderTransport::flush() {
     // Fixups after varint size calculations
-    headerSize = (pkt - headerStart);
+    ptrdiff_t ptrSize = pkt - headerStart;
+	if (ptrSize > std::numeric_limits<uint32_t>().max()) {
```
indentation

hcorg commented on lib/cpp/src/thrift/transport/THeaderTransport.h:
```diff
@@ -135,8 +142,12 @@ class THeaderTransport : public TVirtualTransport<THeaderTransport, TFramedTrans
-    return trans;
+    std::vector<uint16_t>::size_type trans = writeTrans_.size();
+	if (trans > std::numeric_limits<uint16_t>().max()) {
+      throw TTransportException(TTransportException::CORRUPTED_DATA,
```
indent

hcorg commented on lib/cpp/test/JSONProtoTest.cpp:
```diff
@@ -262,8 +262,9 @@ BOOST_AUTO_TEST_CASE(test_json_proto_8) {
   ":[\"i8\",3,1,2,3]},\"13\":{\"lst\":[\"i16\",3,1,2,3]},\"14\":{\"lst\":[\"i64"
   "\",3,1,2,3]}}";
 
+  size_t bufSiz = strlen(json_string)*sizeof(char);
```
`const std::size_t bufSize` could improve readability

jeking3 replied:
Fixed in the next push.

Jens-G commented on compiler/cpp/src/thrift/generate/t_erl_generator.cc:
```diff
@@ -984,10 +988,13 @@ void t_erl_generator::export_types_string(string name, int num) {
 void t_erl_generator::export_function(t_function* tfunction, string prefix) {
-
+  t_struct::members_type::size_type num = tfunction->get_arglist()->get_members().size();
+  if (num > std::numeric_limits<int>().max()) {
```
I'm getting a warning at ` if (num > std::numeric_limits<int>().max()) { ` in this file and also in the erlang generator (2 occurences):

`C4018: ">": Conflict between "signed" and "unsigned"`

jeking3 replied:
Yes, I just finished a suite of batch files to build zlib, openssl, libevent, and thrift for VS 2010 through 2015, win32/x64, debug/release.  I started running into these signed/unsigned issues today and I am working through resolving it.  The acceptance criteria for this issue is that I can build all combinations with zero warnings - we'll see how close I get.

jeking3 commented:
Hi, I pushed fixes for the open issues however I need to make sure it still builds on linux; my local build failed so it looks like there will be another push to clean that up.  Thanks for your patience.

jeking3 commented on lib/cpp/src/thrift/windows/config.h:
```diff
@@ -36,6 +36,9 @@
 
+// Something that defines PRId64 is required to build
+#define HAVE_INTTYPES_H 1
+
```
Note I have opened THRIFT-4025 because I don't see why we have a custom config.h file on Windows when we go through the hassle of generating one with cmake.

hcorg replied:
it's probably left-over from pre-cmake times: vstudio project included in repo did not generate any config.h, yet something had to be included...

jeking3 commented:
There was just one issue on linux; I reverted a change in TBufferTransports.h (I had converted an assert to an exception, but we can figure out what to do with asserts holistically in THRIFT-3978).  Hopefully this will pass.  I have built this on all flavors of Visual Studio 2010, 2012, 2013, 2015 in Debug and Release, Win32 and x64, and run the unit tests.  I also built this on Ubuntu 14.04.  The CI build will prove out any other issues.

jeking3 commented:
Quite the strange error in a couple of the builds.  In the compilation of TNonBlockingServer.cpp it says PRIu32 is unexpected.  The build is locating inttypes.h per the cmake output, so inttypes.h is being included.  I write a small test program and compiled it on Ubuntu 14.04 with gcc 4.6.4 and 4.8.2 and both of them were fine in using that once inttypes.h was included.

jeking3 commented:
The previous CI build failed in a couple ways I cannot explain; one was in building the docker environment and the other was a failure to build something in c_glib in an autoconf build.  I tried to recreate the latter locally using a cmake build environment but I could not.  I rebased to the latest master and did a force push and we'll see what happens.

jeking3 commented:
The remaining two build failures seem unrelated to these changes,.  They are both failures during docker setup.
